File: src/main/java/kr/co/bookand/backend/feedback/domain/dto/FeedbackDto.java
package kr.co.bookand.backend.feedback.domain.dto;

import io.swagger.annotations.ApiModelProperty;
import kr.co.bookand.backend.account.domain.Account;
import kr.co.bookand.backend.feedback.domain.Feedback;
import kr.co.bookand.backend.feedback.domain.FeedbackTarget;
import kr.co.bookand.backend.feedback.domain.FeedbackType;
import lombok.Builder;

public class FeedbackDto {

    public record FeedbackRequest(
            @ApiModelProperty(value = "피드백 유형 (PUSH, INFORMATION_ERROR, INCONVENIENCE, ETC)")
            String feedbackType,

            @ApiModelProperty(value = "피드백 대상 (HOME, MAP, BOOKMARK, MY_PAGE, ETC)")
            String feedbackTarget,
            String content
    ) {
        @Builder
        public FeedbackRequest {
        }

        public Feedback toEntity(Account account) {
            return Feedback.builder()
                    .feedbackType(FeedbackType.valueOf(feedbackType))
                    .feedbackTarget(feedbackTarget == null ? null : FeedbackTarget.valueOf(feedbackTarget))
                    .content(content)
                    .account(account)
                    .build();
        }
    }

    public record FeedbackResponse(
            Long id
    ) {
        public static FeedbackResponse of(Feedback feedback) {
            return new FeedbackResponse(
                    feedback.getId()
            );
        }
    }

    public record FeedbackListResponse(
            Long id,
            String providerEmail,
            String feedbackType,
            String content,
            String createdAt
    ) {
        public static FeedbackListResponse of(Feedback feedback) {
            return new FeedbackListResponse(
                    feedback.getId(),
                    feedback.getAccount().getEmail(),
                    feedback.getFeedbackType().toDetail(),
                    feedback.getContent(),
                    feedback.getCreatedAt()
            );
        }
    }

    public record FeedbackDetailResponse(
            Long id,
            String providerEmail,
            String feedbackType,
            String feedbackTarget,
            String content
    ) {
        public static FeedbackDetailResponse of(Feedback feedback) {
            return new FeedbackDetailResponse(
                    feedback.getId(),
                    feedback.getAccount().getEmail(),
                    feedback.getFeedbackType().toDetail(),
                    feedback.getFeedbackTarget() == null ? null : feedback.getFeedbackTarget().toDetail(),
                    feedback.getContent()
            );
        }
    }
}

File: src/main/kotlin/kr/co/bookand/backend/feedback/dto/FeedbackDto.kt
package kr.co.bookand.backend.feedback.dto

import io.swagger.annotations.ApiModelProperty
import kr.co.bookand.backend.common.DeviceOSFilter
import kr.co.bookand.backend.common.PageResponse
import kr.co.bookand.backend.feedback.model.Feedback

data class CreateFeedbackRequest(
    @ApiModelProperty(value = "피드백 유형 (PUSH, INFORMATION_ERROR, INCONVENIENCE, ETC)")
    val feedbackType: String,
    @ApiModelProperty(value = "피드백 대상 (HOME, MAP, BOOKMARK, MY_PAGE, ARTICLE, BOOKSTORE, ETC)")
    val feedbackTarget: String,
    val content: String,
    val deviceOS: DeviceOSFilter
)

data class FeedbackIdResponse(
    val id: Long
)

data class FeedbackResponse(
    val id: Long,
    val providerEmail: String?,
    val userId: Long,
    val deviceOS: DeviceOSFilter,
    val feedbackType: String,
    val feedbackTarget: String,
    val content: String,
    val createdAt: String,
    val isConfirmed: Boolean
) {
    constructor(feedback: Feedback) : this(
        id = feedback.id,
        providerEmail = feedback.account?.providerEmail,
        userId = feedback.account?.id ?: 0,
        deviceOS = feedback.deviceOS,
        feedbackType = feedback.feedbackType.toDetail(),
        feedbackTarget = feedback.feedbackTarget.toDetail(),
        content = feedback.content,
        createdAt = feedback.createdAt.toString(),
        isConfirmed = feedback.isConfirmed
    )
}

data class FeedbackListResponse(
    val data: PageResponse<FeedbackResponse>,
)